File: agents/negotiator.py
import os
import logging
import json
from typing import Dict, Any, Optional
from dotenv import load_dotenv
from motor.motor_asyncio import AsyncIOMotorCollection
from openai import OpenAI
from models import RequestStatus

# Load environment variables
load_dotenv()

logger = logging.getLogger(__name__)


class NegotiatorAgent:
    def __init__(self):
        self.fireworks_api_key = os.getenv("FIREWORKS_API_KEY")
        if not self.fireworks_api_key:
            raise ValueError("FIREWORKS_API_KEY environment variable is required")
        
        # Fireworks AI uses OpenAI-compatible API
        self.client = OpenAI(
            api_key=self.fireworks_api_key,
            base_url="https://api.fireworks.ai/inference/v1"
        )
        self.model = "accounts/fireworks/models/llama-v3p1-70b-instruct"

    def get_negotiation_functions(self):
        """Define function calling schema for negotiation"""
        return [
            {
                "type": "function",
                "function": {
                    "name": "make_offer",
                    "description": "Make a price offer during negotiation",
                    "parameters": {
                        "type": "object",
                        "properties": {
                            "offer_price": {
                                "type": "number",
                                "description": "The price being offered"
                            },
                            "reasoning": {
                                "type": "string",
                                "description": "Reasoning for this offer"
                            }
                        },
                        "required": ["offer_price", "reasoning"]
                    }
                }
            },
            {
                "type": "function",
                "function": {
                    "name": "accept_offer",
                    "description": "Accept the current offer and finalize the deal",
                    "parameters": {
                        "type": "object",
                        "properties": {
                            "final_price": {
                                "type": "number",
                                "description": "The final agreed price"
                            },
                            "reasoning": {
                                "type": "string",
                                "description": "Reason for accepting"
                            }
                        },
                        "required": ["final_price", "reasoning"]
                    }
                }
            },
            {
                "type": "function",
                "function": {
                    "name": "reject_offer",
                    "description": "Reject the offer and end negotiation",
                    "parameters": {
                        "type": "object",
                        "properties": {
                            "reasoning": {
                                "type": "string",
                                "description": "Reason for rejection"
                            }
                        },
                        "required": ["reasoning"]
                    }
                }
            }
        ]

    async def negotiate(self, budget: float, found_item: Dict[str, Any], negotiation_log: list) -> Dict[str, Any]:
        """Perform negotiation using Fireworks AI function calling"""
        try:
            item_price = found_item.get("price", 0)
            item_name = found_item.get("name", "item")
            
            # Build conversation history from negotiation log
            messages = [
                {
                    "role": "system",
                    "content": f"""You are a skilled negotiator agent. Your goal is to negotiate the best price for an item.
                    
Item Details:
- Name: {item_name}
- Current Price: ${item_price:.2f}
- Your Budget: ${budget:.2f}

Rules:
1. Try to negotiate the price down to fit within the budget
2. Be strategic and reasonable in your offers
3. If the price is already within budget, you can accept it
4. If negotiation fails after reasonable attempts, reject the offer
5. Always use the function calling tools to make offers, accept, or reject

Previous negotiation history:
{json.dumps(negotiation_log[-5:], indent=2) if negotiation_log else "None"}
"""
                },
                {
                    "role": "user",
                    "content": f"Negotiate the price for {item_name}. Current price is ${item_price:.2f} and your budget is ${budget:.2f}."
                }
            ]
            
            # Call Fireworks AI with function calling
            response = self.client.chat.completions.create(
                model=self.model,
                messages=messages,
                tools=self.get_negotiation_functions(),
                tool_choice="auto",
                temperature=0.7,
            )
            
            message = response.choices[0].message
            
            # Handle function calls
            if message.tool_calls:
                tool_call = message.tool_calls[0]
                function_name = tool_call.function.name
                function_args = json.loads(tool_call.function.arguments)
                
                logger.info(f"Negotiation function call: {function_name} with args: {function_args}")
                
                return {
                    "action": function_name,
                    "arguments": function_args,
                    "message": message.content or f"Calling {function_name}",
                }
            else:
                # No function call, treat as message
                return {
                    "action": "message",
                    "message": message.content or "No response",
                }
                
        except Exception as e:
            logger.error(f"Error in negotiation: {e}")
            raise

    async def process_request(self, collection: AsyncIOMotorCollection, request_doc: Dict[str, Any]):
        """Process a FOUND request: negotiate price, update status"""
        request_id = request_doc.get("_id")
        current_status = request_doc.get("status")
        
        try:
            # Only process FOUND requests
            if current_status != RequestStatus.FOUND.value:
                logger.warning(f"Request {request_id} is not in FOUND status, skipping")
                return
            
            budget = request_doc.get("budget", 0)
            found_item = request_doc.get("found_item", {})
            negotiation_log = request_doc.get("negotiation_log", [])
            
            if not found_item:
                logger.warning(f"Request {request_id} has no found_item, skipping")
                return
            
            # Check if already negotiating
            if current_status == RequestStatus.NEGOTIATING.value:
                # Continue negotiation
                pass
            else:
                # Start negotiation - update status to NEGOTIATING
                await collection.update_one(
                    {"_id": request_id},
                    {"$set": {"status": RequestStatus.NEGOTIATING.value}}
                )
                logger.info(f"Request {request_id} status updated to NEGOTIATING")
            
            # Perform negotiation
            negotiation_result = await self.negotiate(budget, found_item, negotiation_log)
            
            # Log the negotiation step
            from datetime import datetime
            log_entry = {
                "timestamp": datetime.utcnow(),
                "agent": "negotiator",
                "message": negotiation_result.get("message", ""),
                "action": negotiation_result.get("action", ""),
            }
            
            # Update based on negotiation result
            action = negotiation_result.get("action")
            args = negotiation_result.get("arguments", {})
            
            if action == "accept_offer":
                final_price = args.get("final_price", found_item.get("price"))
                log_entry["price"] = final_price
                
                # Update status to AGREED
                await collection.update_one(
                    {"_id": request_id},
                    {
                        "$set": {
                            "status": RequestStatus.AGREED.value,
                            "found_item.price": final_price,
                        },
                        "$push": {"negotiation_log": log_entry}
                    }
                )
                logger.info(f"Request {request_id} negotiation accepted at ${final_price:.2f}")
                
            elif action == "reject_offer":
                log_entry["message"] = args.get("reasoning", "Offer rejected")
                
                # Mark as RETRY_SEARCH to try finding another item
                await collection.update_one(
                    {"_id": request_id},
                    {
                        "$set": {"status": RequestStatus.RETRY_SEARCH.value},
                        "$push": {"negotiation_log": log_entry}
                    }
                )
                logger.info(f"Request {request_id} negotiation rejected, marked for retry")
                
            elif action == "make_offer":
                offer_price = args.get("offer_price", 0)
                log_entry["price"] = offer_price
                log_entry["message"] = args.get("reasoning", f"Offering ${offer_price:.2f}")
                
                # Continue negotiation - update log
                await collection.update_one(
                    {"_id": request_id},
                    {"$push": {"negotiation_log": log_entry}}
                )
                logger.info(f"Request {request_id} made offer: ${offer_price:.2f}")
                
            else:
                # Just log the message
                await collection.update_one(
                    {"_id": request_id},
                    {"$push": {"negotiation_log": log_entry}}
                )
                
        except Exception as e:
            logger.error(f"Error processing request {request_id}: {e}")
            raise

File: web/lib/agents/negotiator.ts
import { generateText, tool } from 'ai';
import connectDB from '@/lib/mongodb';
import { Request, RequestStatus } from '@/lib/models/Request';

const FIREWORKS_API_KEY = process.env.FIREWORKS_API_KEY;
const FIREWORKS_API_BASE_URL = process.env.FIREWORKS_API_BASE_URL || 'https://api.fireworks.ai/inference/v1';

if (!FIREWORKS_API_KEY) {
  throw new Error('FIREWORKS_API_KEY environment variable is required');
}

export class NegotiatorAgent {
  private apiKey: string;
  private baseUrl: string;

  constructor() {
    this.apiKey = FIREWORKS_API_KEY;
    this.baseUrl = FIREWORKS_API_BASE_URL;
  }

  /**
   * Process a FOUND request: negotiate with seller
   */
  async processFoundRequest(requestId: string): Promise<void> {
    await connectDB();

    const request = await Request.findById(requestId);
    if (!request) {
      throw new Error(`Request ${requestId} not found`);
    }

    if (request.status !== RequestStatus.FOUND) {
      console.log(`Request ${requestId} is not FOUND, skipping`);
      return;
    }

    if (!request.found_item) {
      throw new Error(`Request ${requestId} has no found_item`);
    }

    try {
      console.log(`Negotiator: Processing FOUND request ${requestId}`);

      // Update status to NEGOTIATING
      request.status = RequestStatus.NEGOTIATING;
      request.negotiation_rounds = 0;
      await request.save();

      const foundItem = request.found_item;
      const budget = request.budget;
      // New schema: use price field, minimum acceptable is 80% of price
      const listedPrice = foundItem.price || 0;
      const minimumPrice = listedPrice * 0.8;

      // Create tools
      const sendOfferToSeller = tool({
        description: 'Send an offer to the seller agent. Returns accept/decline response.',
        parameters: {
          type: 'object',
          properties: {
            item_id: {
              type: 'string',
              description: 'The item ID to make an offer for',
            },
            offer_amount: {
              type: 'number',
              description: 'The offer amount in USDC',
            },
          },
          required: ['item_id', 'offer_amount'],
        },
        execute: async ({ item_id, offer_amount }) => {
          try {
            const baseUrl = process.env.NEXT_PUBLIC_APP_URL || 'http://localhost:3000';
            const response = await fetch(`${baseUrl}/api/seller`, {
              method: 'POST',
              headers: { 'Content-Type': 'application/json' },
              body: JSON.stringify({ item_id, offer_amount }),
            });

            const result = await response.json();
            
            // Log to negotiation_history
            request.negotiation_history.push({
              role: 'negotiator',
              content: `Offered $${offer_amount} for item ${item_id}`,
              offer: offer_amount,
              timestamp: new Date(),
            });

            if (result.accepted) {
              request.negotiation_history.push({
                role: 'seller',
                content: result.message,
                offer: result.accepted_amount,
                timestamp: new Date(),
              });
            } else {
              request.negotiation_history.push({
                role: 'seller',
                content: result.message,
                offer: null,
                timestamp: new Date(),
              });
            }

            await request.save();
            return result;
          } catch (error: any) {
            return {
              accepted: false,
              message: `Error calling seller API: ${error.message}`,
            };
          }
        },
      });

      const processPayment = tool({
        description: 'Process payment using x402 protocol. Call this when seller accepts an offer.',
        parameters: {
          type: 'object',
          properties: {
            offer_id: {
              type: 'string',
              description: 'The offer ID from accepted seller response',
            },
            amount_usdc: {
              type: 'number',
              description: 'The payment amount in USDC',
            },
            buyer_wallet: {
              type: 'string',
              description: 'The buyer wallet address',
            },
            seller_wallet: {
              type: 'string',
              description: 'The seller wallet address from accepted offer',
            },
          },
          required: ['offer_id', 'amount_usdc', 'buyer_wallet', 'seller_wallet'],
        },
        execute: async ({ offer_id, amount_usdc, buyer_wallet, seller_wallet }) => {
          try {
            const baseUrl = process.env.NEXT_PUBLIC_APP_URL || 'http://localhost:3000';
            const response = await fetch(`${baseUrl}/api/quick-pay`, {
              method: 'POST',
              headers: { 'Content-Type': 'application/json' },
              body: JSON.stringify({
                offer_id,
                amount_usdc,
                buyer_wallet,
                seller_wallet,
              }),
            });

            const result = await response.json();
            
            if (result.success) {
              request.status = RequestStatus.PAID;
              request.tx_hash = result.tx_hash;
              await request.save();
            }
            
            return result;
          } catch (error: any) {
            return {
              success: false,
              message: `Error processing payment: ${error.message}`,
            };
          }
        },
      });

      const setRetryStatus = tool({
        description: 'Set request status to RETRY_SEARCH when seller rejects after 3 rounds. Call this when you have exhausted all 3 negotiation rounds.',
        parameters: {
          type: 'object',
          properties: {
            feedback: {
              type: 'string',
              description: 'Feedback message explaining why negotiation failed',
            },
          },
          required: ['feedback'],
        },
        execute: async ({ feedback }) => {
          try {
            request.status = RequestStatus.RETRY_SEARCH;
            request.retry_feedback = feedback;
            request.negotiation_rounds = 0; // Reset for next attempt
            await request.save();

            return {
              success: true,
              message: `Request set to RETRY_SEARCH. Scout will find next alternative item.`,
            };
          } catch (error: any) {
            return {
              success: false,
              message: `Error setting retry status: ${error.message}`,
            };
          }
        },
      });

      // System prompt for negotiation agent
      const systemPrompt = `You are a negotiation agent. You have exactly 3 rounds to negotiate with the seller.
Your goal is to get the seller to accept an offer within the buyer's budget of $${budget}.

Item Details:
- Item ID: ${foundItem.id}
- Size: ${foundItem.size}
- Listed Price: $${listedPrice}
- Minimum Acceptable Price: $${minimumPrice.toFixed(2)}
- Seller Wallet: ${foundItem.seller_wallet}

Rules:
- You can make strategic offers based on the minimum price ($${minimumPrice.toFixed(2)}) and buyer's budget ($${budget})
- Track your negotiation rounds (you have 3 total)
- If seller rejects after 3 rounds, call setRetryStatus() with feedback explaining why negotiation failed
- If seller accepts, call processPayment() immediately with the offer_id and seller_wallet from the response
- Log all interactions to negotiation_history
- Be strategic: start with a reasonable offer, adjust based on seller responses
- Remember: offer_amount must be >= minimum_price ($${minimumPrice})`;

      // Run negotiation loop (max 3 rounds)
      const maxRounds = 3;
      let negotiationRound = 0;
      let messages: Array<{ role: 'user' | 'assistant'; content: string }> = [
        {
          role: 'user',
          content: `Start negotiating for item ${foundItem.id}. Budget: $${budget}, Minimum: $${minimumPrice.toFixed(2)}. You have ${maxRounds} rounds.`,
        },
      ];

      while (negotiationRound < maxRounds) {
        negotiationRound++;
        request.negotiation_rounds = negotiationRound;
        await request.save();

        console.log(`Negotiator: Round ${negotiationRound}/${maxRounds} for request ${requestId}`);

        try {
          const result = await generateText({
            model: {
              provider: 'openai',
              modelId: 'firefunction-v1',
              apiKey: this.apiKey,
              baseURL: this.baseUrl,
            } as any, // Fireworks AI uses OpenAI-compatible API
            tools: {
              sendOfferToSeller,
              processPayment,
              setRetryStatus,
            },
            system: systemPrompt,
            messages,
            maxSteps: 5, // Allow multiple tool calls per round
          });

          // Add assistant response to messages
          messages.push({
            role: 'assistant',
            content: result.text,
          });

          // Check if payment was processed (success)
          if (request.status === RequestStatus.PAID) {
            console.log(`Negotiator: Payment processed successfully for request ${requestId}`);
            return;
          }

          // Check if retry status was set (failure after 3 rounds)
          if (request.status === RequestStatus.RETRY_SEARCH) {
            console.log(`Negotiator: Set retry status for request ${requestId}`);
            return;
          }

          // Check if seller accepted (status should be AGREED)
          if (request.status === RequestStatus.AGREED) {
            // Process payment
            const offerId = request.offer_id;
            if (offerId && foundItem.seller_wallet) {
              const paymentResult = await processPayment.execute({
                offer_id: offerId,
                amount_usdc: request.negotiation_history[request.negotiation_history.length - 1].offer || minimumPrice,
                buyer_wallet: request.buyer_wallet || '0x0000000000000000000000000000000000000000',
                seller_wallet: foundItem.seller_wallet,
              });

              if (paymentResult.success) {
                console.log(`Negotiator: Payment processed successfully!`);
                return;
              }
            }
          }

          // If we're at max rounds and haven't succeeded, set retry status
          if (negotiationRound >= maxRounds && request.status === RequestStatus.NEGOTIATING) {
            await setRetryStatus.execute({
              feedback: `Negotiation failed after ${maxRounds} rounds. Seller rejected all offers.`,
            });
            return;
          }

          // Add user message for next round
          if (negotiationRound < maxRounds) {
            messages.push({
              role: 'user',
              content: `Continue negotiating. Round ${negotiationRound + 1} of ${maxRounds}.`,
            });
          }
        } catch (error: any) {
          console.error(`Error in negotiation round ${negotiationRound}:`, error);
          // Continue to next round or set retry status
          if (negotiationRound >= maxRounds) {
            await setRetryStatus.execute({
              feedback: `Negotiation error: ${error.message}`,
            });
            return;
          }
        }
      }
    } catch (error) {
      console.error(`Error processing FOUND request ${requestId}:`, error);
      request.status = RequestStatus.FAILED;
      request.retry_feedback = `Negotiator error: ${error instanceof Error ? error.message : 'Unknown error'}`;
      await request.save();
      throw error;
    }
  }

  /**
   * Poll for FOUND status requests and process them
   */
  async pollAndProcess(): Promise<void> {
    await connectDB();

    try {
      // Find requests with FOUND status
      const requests = await Request.find({
        status: RequestStatus.FOUND,
      }).limit(5); // Process up to 5 at a time

      console.log(`Negotiator: Found ${requests.length} requests to process`);

      for (const request of requests) {
        try {
          await this.processFoundRequest(request._id.toString());
        } catch (error) {
          console.error(`Error processing request ${request._id}:`, error);
          // Continue processing other requests
        }
      }
    } catch (error) {
      console.error('Error in pollAndProcess:', error);
      throw error;
    }
  }
}
